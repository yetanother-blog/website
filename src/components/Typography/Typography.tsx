import React from 'react';
import styled from 'styled-components';
import {
  variant,
  TypographyProps,
  typography,
  margin,
  MarginBottomProps,
  DisplayProps,
  display,
  color,
} from 'styled-system';
import { StyledComponentPropsWithAs } from '../../types';
import { DefaultTheme } from 'styled-components';

export interface TypographyTypeProps {
  variant: 'title' | 'headline' | 'subheadline' | 'text' | 'smallText';
}

type FormattingTags = 'h1' | 'h2' | 'h3' | 'p' | 'span' | 'em' | 'blockquote' | 'pre';
export type TypoProps = TypographyProps &
  MarginBottomProps &
  TypographyTypeProps &
  DisplayProps &
  StyledComponentPropsWithAs<FormattingTags, DefaultTheme, {}, any>;

const variants = variant({
  variants: {
    title: {
      fontSize: [32, 42],
      fontWeight: 600,
      letterSpacing: 0.5,
      lineHeight: 1.25,
      marginBottom: 0,
    },
    headline: {
      fontSize: [26, 30],
      lineHeight: 1.25,
      marginBottom: 0,
    },
    subheadline: {
      fontSize: 25,
      lineHeight: 1.3,
      marginBottom: 0,
    },
    text: {
      fontSize: 20,
      marginBottom: 0,
    },
    smallText: {
      fontSize: 16,
      lineHeight: 1.4,
      marginBottom: 0,
    },
  },
});

const StyledTitle = styled.h1`
  -webkit-font-smoothing: antialiased;
  ${display}
  ${color}
  ${typography}
  ${variants}
  ${margin}
`;

const StyledHeadline = styled.h2`
  -webkit-font-smoothing: antialiased;
  ${display}
  ${color}
  ${typography}
  ${variants}
  ${margin}
`;

const StyledSubheadline = styled.h3`
  -webkit-font-smoothing: antialiased;
  ${display}
  ${color}
  ${typography}
  ${variants}
  ${margin}
`;

const StyledText = styled.p`
  -webkit-font-smoothing: antialiased;
  ${display}
  ${color}
  ${typography}
  ${variants}
  ${margin}
`;

const StyledSmallText = styled.p`
  -webkit-font-smoothing: antialiased;
  ${display}
  ${color}
  ${typography}
  ${variants}
  ${margin}
`;

export const Typography: React.FC<TypoProps> = (props) => {
  switch (props.variant) {
    case 'title': {
      return <StyledTitle {...props} />;
    }
    case 'headline': {
      return <StyledHeadline {...props} />;
    }
    case 'subheadline': {
      return <StyledSubheadline {...props} />;
    }
    case 'text': {
      return <StyledText {...props} />;
    }
    default: {
      return <StyledSmallText {...props} />;
    }
  }
};
