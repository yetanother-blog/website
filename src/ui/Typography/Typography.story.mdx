import { Story, Meta, Preview, Props } from '@storybook/addon-docs/blocks';
import { Typography } from './Typography';
import { ThemeContext } from 'styled-components';

<Meta title="Components|Typography" />

# Typography

Typogrpahy component to wrap pages and posts.

<Preview>
  <Story name="default">
    <ThemeContext.Consumer>
      {(theme) => (
        <React.Fragment>
          <Typography variant="title" marginBottom={theme.space.xl}>
            Hallo Typography
          </Typography>
          <Typography variant="headline" marginBottom={theme.space.xl}>
            Hallo Typography
          </Typography>
          <Typography variant="subheadline" marginBottom={theme.space.xl}>
            Hallo Typography
          </Typography>
          <Typography variant="text" marginBottom={theme.space.xl}>
            Hallo Typography
          </Typography>
          <Typography variant="smallText" marginBottom={theme.space.xl}>
            Hallo Typography
          </Typography>
        </React.Fragment>
      )}
    </ThemeContext.Consumer>
  </Story>
</Preview>

## Props

<Props of={Typography} />
