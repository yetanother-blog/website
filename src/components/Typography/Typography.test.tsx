import React from 'react';
import { render } from '@testing-library/react';
import { Typography } from './Typography';

describe('Typography component', () => {
  it('should render a box w/ content', () => {
    const { container } = render(<Typography variant="title">Hello Typography ðŸŽ‰</Typography>);

    expect(container).toMatchSnapshot();
  });
});
