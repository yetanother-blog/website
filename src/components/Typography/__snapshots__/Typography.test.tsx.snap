// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography component should render a box w/ content 1`] = `
<div>
  <h1
    class="sc-bdvvaa QPUsR"
  >
    Hello Typography ðŸŽ‰
  </h1>
</div>
`;
